File: Android/virtues_dynamic/src/main/java/com/appdimens/dynamic/compose/AppDimensFixed.kt
/**
 * Author & Developer: Jean Bodenberg
 * GIT: https://github.com/www-virtues-ag/virtues-dimens.git
 * Date: 2025-10-04
 *
 * Library: Virtues
 *
 * Description:
 * The Virtues library is a dimension management system that automatically
 * adjusts Dp, Sp, and Px values in a responsive and mathematically refined way,
 * ensuring layout consistency across any screen size or ratio.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ag.virtues.dimens.dynamic.compose

import android.annotation.SuppressLint
import android.content.res.Configuration
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import ag.virtues.dimens.dynamic.compose.VirtuesAdjustmentFactors.BASE_DP_FACTOR
import ag.virtues.dimens.dynamic.compose.VirtuesAdjustmentFactors.BASE_INCREMENT
import ag.virtues.dimens.dynamic.compose.VirtuesAdjustmentFactors.REFERENCE_AR
import ag.virtues.dimens.dynamic.compose.VirtuesAdjustmentFactors.getReferenceAspectRatio
import ag.virtues.dimens.dynamic.compose.VirtuesAdjustmentFactors.rememberAdjustmentFactors
import ag.virtues.dimens.dynamic.compose.VirtuesAdjustmentFactors.resolveIntersectionCondition
import ag.virtues.dimens.dynamic.compose.VirtuesAdjustmentFactors.resolveQualifierDp
import ag.virtues.dimens.library.DpQualifier
import ag.virtues.dimens.library.DpQualifierEntry
import ag.virtues.dimens.library.ScreenType
import ag.virtues.dimens.library.UiModeQualifierEntry
import ag.virtues.dimens.library.UiModeType
import kotlin.math.ln

/**
 * [EN] A builder class for creating "fixed" dimensions that are automatically adjusted
 * based on the device's `smallestScreenWidthDp` and screen aspect ratio.
 *
 * [PT] Uma classe construtora para criar dimensões "fixas" que são ajustadas
 * automaticamente com base no `smallestScreenWidthDp` do dispositivo e na proporção da tela.
 */
class VirtuesFixed(
    private val initialBaseDp: Dp,
    private var ignoreMultiViewAdjustment: Boolean = false
) {
    /**
     * [EN] Map to store custom Dp values (Priority 3).
     *
     * [PT] Mapa para armazenar valores Dp customizados (Prioridade 3).
     */
    private var customDpMap: MutableMap<DpQualifierEntry, Dp> = mutableMapOf()

    /**
     * [EN] Map for custom Dp values by UiModeType (Priority 2).
     *
     * [PT] Mapa para valores Dp customizados por UiModeType (Prioridade 2).
     */
    private var customUiModeMap: MutableMap<UiModeType, Dp> = mutableMapOf()

    /**
     * [EN] Map for custom Dp values by INTERSECTION (UiMode + DpQualifier) (Priority 1).
     *
     * [PT] Mapa para valores Dp customizados por INTERSEÇÃO (UiMode + DpQualifier) (Prioridade 1).
     */
    private var customIntersectionMap: MutableMap<UiModeQualifierEntry, Dp> = mutableMapOf()

    /**
     * [EN] Indicates whether aspect ratio-based adjustment should be applied.
     *
     * [PT] Indica se o ajuste baseado na proporção (aspect ratio) deve ser aplicado.
     */
    private var applyAspectRatioAdjustment: Boolean = true

    /**
     * [EN] Custom sensitivity; uses the default value if null.
     *
     * [PT] Sensibilidade customizada; usa o valor padrão se for nulo.
     */
    private var customSensitivityK: Float? = null

    /**
     * [EN] Defines the screen dimension to be used as the base (HIGHEST or LOWEST).
     *
     * [PT] Define a dimensão da tela a ser usada como base (HIGHEST ou LOWEST).
     */
    private var screenType: ScreenType = ScreenType.LOWEST

    /**
     * [EN] Sets a custom dimension for a specific UI mode.
     *
     * [PT] Define uma dimensão customizada para um modo de UI específico.
     */
    fun screen(type: UiModeType, customValue: Dp): VirtuesFixed {
        customUiModeMap[type] = customValue
        return this
    }

    /**
     * [EN] Overload for `screen` that accepts a TextUnit value.
     *
     * [PT] Sobrecarga para `screen` que aceita um valor TextUnit.
     */
    fun screen(type: UiModeType, customValue: TextUnit): VirtuesFixed {
        customUiModeMap[type] = customValue.value.dp
        return this
    }

    /**
     * [EN] Overload for `screen` that accepts a Float value.
     *
     * [PT] Sobrecarga para `screen` que aceita um valor Float.
     */
    fun screen(type: UiModeType, customValue: Float): VirtuesFixed {
        customUiModeMap[type] = customValue.dp
        return this
    }

    /**
     * [EN] Overload for `screen` that accepts an Int value.
     *
     * [PT] Sobrecarga para `screen` que aceita um valor Int.
     */
    fun screen(type: UiModeType, customValue: Int): VirtuesFixed {
        customUiModeMap[type] = customValue.dp
        return this
    }

    /**
     * [EN] Sets a custom dimension for a specific intersection of UI mode and screen qualifier.
     *
     * [PT] Define uma dimensão customizada para uma interseção específica de modo de UI e qualificador de tela.
     */
    fun screen(
        uiModeType: UiModeType, qualifierType: DpQualifier, qualifierValue: Int, customValue: Dp
    ): VirtuesFixed {
        val key = UiModeQualifierEntry(
            uiModeType = uiModeType,
            dpQualifierEntry = DpQualifierEntry(qualifierType, qualifierValue)
        )
        customIntersectionMap[key] = customValue
        return this
    }

    /**
     * [EN] Overload for `screen` intersection that accepts a TextUnit value.
     *
     * [PT] Sobrecarga para a interseção `screen` que aceita um valor TextUnit.
     */
    fun screen(
        uiModeType: UiModeType,
        qualifierType: DpQualifier,
        qualifierValue: Int,
        customValue: TextUnit
    ): VirtuesFixed {
        val key = UiModeQualifierEntry(
            uiModeType = uiModeType,
            dpQualifierEntry = DpQualifierEntry(qualifierType, qualifierValue)
        )
        customIntersectionMap[key] = customValue.value.dp
        return this
    }

    /**
     * [EN] Overload for `screen` intersection that accepts a Float value.
     *
     * [PT] Sobrecarga para a interseção `screen` que aceita um valor Float.
     */
    fun screen(
        uiModeType: UiModeType,
        qualifierType: DpQualifier,
        qualifierValue: Int,
        customValue: Float
    ): VirtuesFixed {
        val key = UiModeQualifierEntry(
            uiModeType = uiModeType,
            dpQualifierEntry = DpQualifierEntry(qualifierType, qualifierValue)
        )
        customIntersectionMap[key] = customValue.dp
        return this
    }

    /**
     * [EN] Overload for `screen` intersection that accepts an Int value.
     *
     * [PT] Sobrecarga para a interseção `screen` que aceita um valor Int.
     */
    fun screen(
        uiModeType: UiModeType,
        qualifierType: DpQualifier,
        qualifierValue: Int,
        customValue: Int
    ): VirtuesFixed {
        val key = UiModeQualifierEntry(
            uiModeType = uiModeType,
            dpQualifierEntry = DpQualifierEntry(qualifierType, qualifierValue)
        )
        customIntersectionMap[key] = customValue.dp
        return this
    }

    /**
     * [EN] Sets a custom dimension for a specific screen qualifier.
     *
     * [PT] Define uma dimensão customizada para um qualificador de tela específico.
     */
    fun screen(type: DpQualifier, value: Int, customValue: Dp): VirtuesFixed {
        customDpMap[DpQualifierEntry(type, value)] = customValue
        return this
    }

    /**
     * [EN] Overload for `screen` that accepts a TextUnit value for a DpQualifier.
     *
     * [PT] Sobrecarga para `screen` que aceita um valor TextUnit para um DpQualifier.
     */
    fun screen(type: DpQualifier, value: Int, customValue: TextUnit): VirtuesFixed {
        customDpMap[DpQualifierEntry(type, value)] = customValue.value.dp
        return this
    }

    /**
     * [EN] Overload for `screen` that accepts a Float value for a DpQualifier.
     *
     * [PT] Sobrecarga para `screen` que aceita um valor Float para um DpQualifier.
     */
    fun screen(type: DpQualifier, value: Int, customValue: Float): VirtuesFixed {
        customDpMap[DpQualifierEntry(type, value)] = customValue.dp
        return this
    }

    /**
     * [EN] Overload for `screen` that accepts an Int value for a DpQualifier.
     *
     * [PT] Sobrecarga para `screen` que aceita um valor Int para um DpQualifier.
     */
    fun screen(type: DpQualifier, value: Int, customValue: Int): VirtuesFixed {
        customDpMap[DpQualifierEntry(type, value)] = customValue.dp
        return this
    }

    /**
     * [EN] Enables or disables aspect ratio adjustment and sets a custom sensitivity.
     *
     * [PT] Ativa ou desativa o ajuste de proporção da tela e define uma sensibilidade customizada.
     */
    fun aspectRatio(enable: Boolean = true, sensitivityK: Float? = null): VirtuesFixed {
        applyAspectRatioAdjustment = enable
        customSensitivityK = sensitivityK
        return this
    }

    /**
     * [EN] Sets the screen dimension type to be used as a reference (HIGHEST or LOWEST).
     *
     * [PT] Define o tipo de dimensão da tela a ser usado como referência (HIGHEST ou LOWEST).
     */
    fun type(type: ScreenType): VirtuesFixed {
        screenType = type
        return this
    }

    /**
     * [EN] Ignores multi-view adjustment if set to true.
     *
     * [PT] Ignora o ajuste de multi-view se definido como verdadeiro.
     */
    fun multiViewAdjustment(ignore: Boolean = true): VirtuesFixed {
        ignoreMultiViewAdjustment = ignore
        return this
    }

    @SuppressLint("ConfigurationScreenWidthHeight")
    @Composable
    private fun rememberFinalBaseDp(): Dp {
        val configuration = LocalConfiguration.current

        val smallestWidthDp = configuration.smallestScreenWidthDp.toFloat()
        val currentScreenWidthDp = configuration.screenWidthDp.toFloat()
        val currentScreenHeightDp = configuration.screenHeightDp.toFloat()

        val currentUiModeType = UiModeType.fromConfiguration(configuration.uiMode)

        return remember(
            smallestWidthDp, currentScreenWidthDp, currentScreenHeightDp,
            customDpMap.hashCode(), customUiModeMap.hashCode(), customIntersectionMap.hashCode(),
            currentUiModeType
        ) {
            var dpToAdjust = initialBaseDp
            var foundCustomDp: Dp?

            /**
             * [EN] PRIORITY 1: INTERSECTION (UiMode + DpQualifier)
             *
             * [PT] PRIORIDADE 1: INTERSEÇÃO (UiMode + DpQualifier)
             */
            val sortedIntersectionQualifiers = customIntersectionMap.entries.toList()
                .sortedByDescending { it.key.dpQualifierEntry.value }

            foundCustomDp = sortedIntersectionQualifiers.firstOrNull { (key, _) ->
                key.uiModeType == currentUiModeType && resolveIntersectionCondition(
                    entry = key.dpQualifierEntry,
                    smallestWidthDp = smallestWidthDp,
                    currentScreenWidthDp = currentScreenWidthDp,
                    currentScreenHeightDp = currentScreenHeightDp
                )
            }?.value

            if (foundCustomDp != null) {
                dpToAdjust = foundCustomDp
            } else {
                /**
                 * [EN] PRIORITY 2: UI MODE (UiModeType only)
                 *
                 * [PT] PRIORIDADE 2: UI MODE (Apenas UiModeType)
                 */
                foundCustomDp = customUiModeMap[currentUiModeType]

                if (foundCustomDp != null) {
                    dpToAdjust = foundCustomDp
                } else {
                    /**
                     * [EN] PRIORITY 3: DP QUALIFIER (SW, H, W only)
                     *
                     * [PT] PRIORIDADE 3: DP QUALIFIER (Apenas SW, H, W)
                     */
                    dpToAdjust = resolveQualifierDp(
                        customDpMap = customDpMap,
                        smallestWidthDp = smallestWidthDp,
                        currentScreenWidthDp = currentScreenWidthDp,
                        currentScreenHeightDp = currentScreenHeightDp,
                        initialBaseDp = initialBaseDp
                    )
                }
            }
            dpToAdjust
        }
    }

    @SuppressLint("ConfigurationScreenWidthHeight")
    @Composable
    private fun calculate(): Float {
        val dpToAdjust = rememberFinalBaseDp()
        val adjustmentFactors = rememberAdjustmentFactors()
        val configuration = LocalConfiguration.current
        var isMultiWindow = ignoreMultiViewAdjustment

        if (ignoreMultiViewAdjustment) {
            val smallestWidthDp = configuration.smallestScreenWidthDp.toFloat()
            val currentScreenWidthDp = configuration.screenWidthDp.toFloat()
            val isLayoutSplit = configuration.screenLayout and Configuration.SCREENLAYOUT_SIZE_MASK != Configuration.SCREENLAYOUT_SIZE_MASK
            val isSmallDifference = (smallestWidthDp - currentScreenWidthDp) < (smallestWidthDp * 0.1)
            isMultiWindow = isLayoutSplit && !isSmallDifference
        }

        val shouldIgnoreAdjustment = ignoreMultiViewAdjustment && isMultiWindow

        val finalAdjustmentFactor = if (shouldIgnoreAdjustment) {
            BASE_DP_FACTOR
        } else if (applyAspectRatioAdjustment) {

            val selectedFactor = when (screenType) {
                ScreenType.HIGHEST -> adjustmentFactors.withArFactorHighest
                ScreenType.LOWEST -> adjustmentFactors.withArFactorLowest
            }

            if (customSensitivityK != null) {
                /**
                 * [EN] Recalculation of the increment factor with custom sensitivity.
                 *
                 * [PT] Recálculo do fator de incremento com sensibilidade customizada.
                 */
                val adjustmentFactorBase = when (screenType) {
                    ScreenType.HIGHEST -> adjustmentFactors.adjustmentFactorHighest
                    ScreenType.LOWEST -> adjustmentFactors.adjustmentFactorLowest
                }

                val currentScreenWidthDp = configuration.screenWidthDp.toFloat()
                val currentScreenHeightDp = configuration.screenHeightDp.toFloat()
                val ar = getReferenceAspectRatio(currentScreenWidthDp, currentScreenHeightDp)
                val continuousAdjustment = (customSensitivityK!! * ln(ar / REFERENCE_AR))
                val finalIncrementValue = BASE_INCREMENT + continuousAdjustment

                BASE_DP_FACTOR + adjustmentFactorBase * finalIncrementValue

            } else {
                selectedFactor /** [EN] Use the memoized factor. [PT] Usa o fator memorizado. */
            }

        } else {
            adjustmentFactors.withoutArFactor /** [EN] No AR. [PT] Sem AR. */
        }

        return dpToAdjust.value * finalAdjustmentFactor
    }

    /**
     * [EN] Builds the adjusted Dp from the calculation.
     *
     * [PT] Constrói o Dp ajustado a partir do cálculo.
     */
    @get:Composable
    val dp: Dp
        get() = calculate().dp

    /**
     * [EN] Builds the adjusted TextUnit (Sp) from the calculation.
     *
     * [PT] Constrói o TextUnit (Sp) ajustado a partir do cálculo.
     */
    @get:Composable
    val sp: TextUnit
        get() = calculate().sp

    /**
     * [EN] Builds the adjusted TextUnit (Sp) from the calculation (NO FONT SCALE).
     *
     * [PT] Constrói o TextUnit (Sp) ajustado a partir do cálculo (SEM ESCALA DE FONTE).
     */
    @get:Composable
    val em: TextUnit
        get() = (calculate() / LocalDensity.current.fontScale).sp

    /**
     * [EN] Builds the adjusted Pixel value (Float) from the calculation.
     *
     * [PT] Constrói o valor em Pixels (Float) ajustado a partir do cálculo.
     */
    @get:Composable
    val px: Float
        get() = calculate()
}

File: iOS/Sources/VirtuesDimensCore/AppDimensFixed.swift
/**
 * Author & Developer: Jean Bodenberg
 * GIT: https://github.com/www-virtues-ag/appdimens.git
 * Date: 2025-01-15
 *
 * Library: VirtuesDimens iOS
 *
 * Description:
 * The VirtuesDimens library is a dimension management system that automatically
 * adjusts Dp, Sp, and Px values in a responsive and mathematically refined way,
 * ensuring layout consistency across any screen size or ratio.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation
import UIKit

/**
 * [EN] Class for building "fixed" dimensions that are automatically adjusted
 * based on the device's screen dimensions and aspect ratio.
 * Compatible with both UIKit and SwiftUI.
 * [PT] Classe para construir dimensões "fixas" que são ajustadas automaticamente
 * com base nas dimensões da tela do dispositivo e na proporção da tela.
 * Compatível com UIKit e SwiftUI.
 */
public class VirtuesDimensFixed {
    
    // MARK: - Properties
    
    private let initialBaseValue: CGFloat
    private var ignoreMultiWindowAdjustment: Bool = false
    private var applyAspectRatioAdjustment: Bool = true
    private var customSensitivityK: CGFloat? = nil
    private var screenType: ScreenType = .lowest
    
    /// [EN] Map to store custom values based on device type (Priority 2).
    /// [PT] Mapa para armazenar valores customizados com base no tipo de dispositivo (Prioridade 2).
    private var customDeviceTypeMap: [DeviceType: CGFloat] = [:]
    
    /// [EN] Map to store custom values based on screen qualifier (Priority 3).
    /// [PT] Mapa para armazenar valores customizados com base no qualificador de tela (Prioridade 3).
    private var customScreenQualifierMap: [ScreenQualifierEntry: CGFloat] = [:]
    
    // MARK: - Initialization
    
    /**
     * [EN] Initializes the VirtuesDimensFixed with a base value.
     * [PT] Inicializa o VirtuesDimensFixed com um valor base.
     */
    public init(_ initialValue: CGFloat, ignoreMultiWindowAdjustment: Bool = false) {
        self.initialBaseValue = initialValue
        self.ignoreMultiWindowAdjustment = ignoreMultiWindowAdjustment
    }
    
    // MARK: - Configuration Methods
    
    /**
     * [EN] Sets a custom dimension value for a specific device type.
     * @param type The device type.
     * @param customValue The custom dimension value in points.
     * @return The VirtuesDimensFixed instance for chaining.
     * [PT] Define um valor de dimensão customizado para um tipo de dispositivo específico.
     * @param type O tipo de dispositivo.
     * @param customValue O valor de dimensão customizado em pontos.
     * @return A instância VirtuesDimensFixed para encadeamento.
     */
    @discardableResult
    public func screen(_ type: DeviceType, _ customValue: CGFloat) -> VirtuesDimensFixed {
        customDeviceTypeMap[type] = customValue
        return self
    }
    
    /**
     * [EN] Sets a custom dimension value for a specific device type and screen size.
     * @param deviceType The device type.
     * @param screenSize The minimum screen size for this qualifier.
     * @param customValue The custom dimension value in points.
     * @return The VirtuesDimensFixed instance for chaining.
     * [PT] Define um valor de dimensão customizado para um tipo de dispositivo e tamanho de tela específicos.
     * @param deviceType O tipo de dispositivo.
     * @param screenSize O tamanho mínimo de tela para este qualificador.
     * @param customValue O valor de dimensão customizado em pontos.
     * @return A instância VirtuesDimensFixed para encadeamento.
     */
    @discardableResult
    public func screen(_ deviceType: DeviceType, _ screenSize: CGFloat, _ customValue: CGFloat) -> VirtuesDimensFixed {
        let entry = ScreenQualifierEntry(deviceType: deviceType, screenSize: screenSize)
        customScreenQualifierMap[entry] = customValue
        return self
    }
    
    /**
     * [EN] Enables or disables the aspect ratio adjustment.
     * @param enable If true, enables the adjustment.
     * @param sensitivity Optional custom sensitivity for the adjustment.
     * @return The VirtuesDimensFixed instance for chaining.
     * [PT] Ativa ou desativa o ajuste de proporção.
     * @param enable Se verdadeiro, ativa o ajuste.
     * @param sensitivity Sensibilidade customizada opcional para o ajuste.
     * @return A instância VirtuesDimensFixed para encadeamento.
     */
    @discardableResult
    public func aspectRatio(enable: Bool = true, sensitivity: CGFloat? = nil) -> VirtuesDimensFixed {
        applyAspectRatioAdjustment = enable
        customSensitivityK = sensitivity
        return self
    }
    
    /**
     * [EN] Sets the screen dimension type (LOWEST or HIGHEST) to be used as the base for adjustments.
     * @param type The screen dimension type.
     * @return The VirtuesDimensFixed instance for chaining.
     * [PT] Define o tipo de dimensão da tela (LOWEST ou HIGHEST) a ser usado como base para os ajustes.
     * @param type O tipo de dimensão da tela.
     * @return A instância VirtuesDimensFixed para encadeamento.
     */
    @discardableResult
    public func type(_ type: ScreenType) -> VirtuesDimensFixed {
        screenType = type
        return self
    }
    
    /**
     * [EN] Ignores adjustments when the app is in multi-window mode.
     * @param ignore If true, adjustments are ignored in multi-window mode.
     * @return The VirtuesDimensFixed instance for chaining.
     * [PT] Ignora os ajustes quando o aplicativo está em modo multi-janela.
     * @param ignore Se verdadeiro, os ajustes são ignorados no modo multi-janela.
     * @return A instância VirtuesDimensFixed para encadeamento.
     */
    @discardableResult
    public func multiWindowAdjustment(ignore: Bool = true) -> VirtuesDimensFixed {
        ignoreMultiWindowAdjustment = ignore
        return self
    }
    
    // MARK: - Value Resolution
    
    /**
     * [EN] Resolves the base value to be adjusted by applying the customization logic
     * (Device Type > Screen Qualifier > Initial Value).
     * [PT] Resolve o valor base a ser ajustado, aplicando a lógica de customização
     * (Tipo de Dispositivo > Qualificador de Tela > Valor Inicial).
     */
    private func resolveFinalBaseValue() -> CGFloat {
        let currentDeviceType = DeviceType.current()
        let (screenWidth, screenHeight) = VirtuesDimensAdjustmentFactors.getCurrentScreenDimensions()
        let currentScreenSize = min(screenWidth, screenHeight)
        
        // Priority 1: Device Type
        if let deviceTypeValue = customDeviceTypeMap[currentDeviceType] {
            return deviceTypeValue
        }
        
        // Priority 2: Screen Qualifier
        let qualifierValue = VirtuesDimensAdjustmentFactors.resolveQualifierValue(
            customMap: customScreenQualifierMap,
            currentDeviceType: currentDeviceType,
            currentScreenSize: currentScreenSize,
            initialValue: initialBaseValue
        )
        
        return qualifierValue
    }
    
    /**
     * [EN] Performs the final dimension adjustment calculation.
     * @return The adjusted value as a CGFloat (not converted to pixels).
     * [PT] Executa o cálculo final do ajuste da dimensão.
     * @return O valor ajustado como CGFloat (não convertido para pixels).
     */
    public func calculateAdjustedValue() -> CGFloat {
        let valueToAdjust = resolveFinalBaseValue()
        let adjustmentFactors = VirtuesDimensAdjustmentFactors.calculateAdjustmentFactors()
        
        let shouldIgnoreAdjustment = ignoreMultiWindowAdjustment && VirtuesDimensAdjustmentFactors.isMultiWindowMode()
        
        let finalAdjustmentFactor: CGFloat
        
        if shouldIgnoreAdjustment {
            finalAdjustmentFactor = VirtuesDimensAdjustmentFactors.BASE_DP_FACTOR
        } else if applyAspectRatioAdjustment {
            let selectedFactor = screenType == .highest ? 
                adjustmentFactors.withArFactorHighest : 
                adjustmentFactors.withArFactorLowest
            
            if let customSensitivity = customSensitivityK {
                let adjustmentFactorBase = screenType == .highest ? 
                    adjustmentFactors.adjustmentFactorHighest : 
                    adjustmentFactors.adjustmentFactorLowest
                
                let (screenWidth, screenHeight) = VirtuesDimensAdjustmentFactors.getCurrentScreenDimensions()
                let ar = VirtuesDimensAdjustmentFactors.getReferenceAspectRatio(screenWidth, screenHeight)
                let continuousAdjustment = customSensitivity * log(ar / VirtuesDimensAdjustmentFactors.REFERENCE_AR)
                let finalIncrementValue = VirtuesDimensAdjustmentFactors.BASE_INCREMENT + continuousAdjustment
                
                finalAdjustmentFactor = VirtuesDimensAdjustmentFactors.BASE_DP_FACTOR + adjustmentFactorBase * finalIncrementValue
            } else {
                finalAdjustmentFactor = selectedFactor
            }
        } else {
            finalAdjustmentFactor = adjustmentFactors.withoutArFactor
        }
        
        return valueToAdjust * finalAdjustmentFactor
    }
    
    // MARK: - Conversion Methods
    
    /**
     * [EN] Returns the adjusted value in points.
     * @return The value in points as a CGFloat.
     * [PT] Retorna o valor ajustado em pontos.
     * @return O valor em pontos como CGFloat.
     */
    public func toPoints() -> CGFloat {
        return calculateAdjustedValue()
    }
    
    /**
     * [EN] Returns the adjusted value in pixels.
     * @return The value in pixels as a CGFloat.
     * [PT] Retorna o valor ajustado em pixels.
     * @return O valor em pixels como CGFloat.
     */
    public func toPixels() -> CGFloat {
        let adjustedValue = calculateAdjustedValue()
        return VirtuesDimensAdjustmentFactors.pointsToPixels(adjustedValue)
    }
    
    /**
     * [EN] Returns the adjusted value as an integer in points.
     * @return The value in points as an Int.
     * [PT] Retorna o valor ajustado como inteiro em pontos.
     * @return O valor em pontos como Int.
     */
    public func toPointsInt() -> Int {
        return Int(calculateAdjustedValue())
    }
    
    /**
     * [EN] Returns the adjusted value as an integer in pixels.
     * @return The value in pixels as an Int.
     * [PT] Retorna o valor ajustado como inteiro em pixels.
     * @return O valor em pixels como Int.
     */
    public func toPixelsInt() -> Int {
        return Int(toPixels())
    }
}
